### PR TITLE
new: stop_manager_plugin (to manage STOP requests & alike)

### PR DESCRIPTION
**Pull request by tenzap**
To be in line with legal requirements in some countries
Add the ability for the user to reply STOP to any SMS to opt-out from future SMS

### CODE REVIEW

**Comment by kingster** [comment]
Shouldn't this rather reuse `blacklist_number` plugin to implement the feature ?

**Comment by tenzap** [comment]
It doesn't have the same kind of feature so I don't think.

Blacklist is "Autoremove incoming SMS from Blacklist number", so if I understand well, it removes the message from the inbox, like a spam filter and avoids sending SMS to someone in the blacklist. The blacklist has to be filled manually. So this is a feature for the kalkun user and requires manual input of numbers in blacklist.

Stop_manager is to allow SMS recipient to opt-out/opt-in from future SMS, so filling of the table is "recipient driven" : 
- detects if the sender asked for Opt-out (by configurable keyword), and one can also fine tune opt-out to specific kind of "mailing-lists" (eg, recipient sends back as SMS containing "STOP")
- also manages opt-in (depending on chosen config)
- detects if outgoing message should be sent or not (depending on config of stop manager & presence in the stop_manager table)
- this can all be configured (keywords, what feature to enable)

**Comment by kingster** [comment]
Sorry I meant `Whitelist Number` plugin also offers similar feature (though manually configured). Can we not combine this two and allow user to choose whichever mode the user wants ?

**Comment by tenzap** [comment]
I'm sorry I still don't get it what is common. Whitelist is just the contrary no ? ie. Don't send if not in whitelist. Still not the same as stop_manager

User may use the plugin he prefers according to his needs

File: application/config/kalkun_plugin_stop_manager.php [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    $config['enable_autoreply_info'] = TRUE;
+    // Send autoreply to tell command is invalid
+    $config['enable_autoreply_error'] = TRUE;
+    // Enable opt-in
+    $config['enable_optin'] = TRUE;
+    // Enable the use of "type" in optin/optout
+    $config['enable_type'] = TRUE;
+
+    // Send only if this is a french mobile phone number ( +336 et +337 )
+    $config['autoreply_outnumber_match_rule'] = '/\+33[67][0-9]{8}/';
```
Comment (kingster) [inline application/config/kalkun_plugin_stop_manager.php:25]: there should be a flag to enable/disable this rule.

File: application/plugins/stop_manager/controllers/Stop_manager.php [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+        $this->load->library('pagination');
+        $config['base_url'] = site_url('plugin/stop_manager');
+        $config['total_rows'] = $this->Stop_manager_model->get('count');
+        $config['per_page'] = $this->Kalkun_model->get_setting()->row('paging');
+        $config['cur_tag_open'] = '<span id="current">';
+        $config['cur_tag_close'] = '</span>';
+        ($this->uri->segment(3,0) == "index") ? $config['uri_segment'] = 4 : $config['uri_segment'] = 3;
+        $this->pagination->initialize($config);
+
+        $offset = ($this->uri->segment(3,0) == "index") ? $this->uri->segment(4,0) : $this->uri->segment(3,0);
```
Comment (kingster) [inline application/plugins/stop_manager/controllers/Stop_manager.php:52]: Is there any better way to access these parameters?

File: application/plugins/stop_manager/media/mysql_stop_manager.sql [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+--
+
+CREATE TABLE IF NOT EXISTS `plugin_stop_manager` (
+    `id_stop_manager` int(5) NOT NULL AUTO_INCREMENT,
+    `destination_number` varchar(20) NOT NULL,
+    `stop_type` varchar(50) NOT NULL,
+    `stop_message` varchar(2000) NOT NULL,
+    `reg_date` datetime NOT NULL,
+    PRIMARY KEY (`id_stop_manager`)
+) ENGINE=MyISAM  DEFAULT CHARSET=latin1;
```
Comment (kingster) [inline application/plugins/stop_manager/media/mysql_stop_manager.sql:14]: please remove `ENGINE=MyISAM` and change charset to utf8

File: application/plugins/stop_manager/stop_manager.php [hunk 4]
```diff
@@ -0,0 +1,268 @@
...
+                    autoreply($from, $text);
+                break;
+            default:
+                $text = "Demande non valide ($msg)";
+                if ($config['enable_autoreply_error'])
+                    autoreply($from, $text);
+                break;
+        }
+    } else {
+        $strTemplate = "Demande non valide (:received_command). Répondre ':optout_keyword";
```
Comment (kingster) [inline application/plugins/stop_manager/stop_manager.php:219]: please move localisation to its respective files.

File: application/plugins/stop_manager/controllers/Stop_manager.php (hunk 2 above)
Comment (tenzap) [reply application/plugins/stop_manager/controllers/Stop_manager.php:52]: I mostly copied what is in other parts of kalkun. I guess there isn't

File: application/plugins/stop_manager/stop_manager.php (hunk 4 above)
Comment (tenzap) [reply application/plugins/stop_manager/stop_manager.php:219]: What does "tni_" stand for in the kalkun_lang.php files ? I'm looking for a place to put the translations and was wondering is there is anything specific to plugins

File: application/config/kalkun_plugin_stop_manager.php (hunk 1 above)
Comment (tenzap) [reply application/config/kalkun_plugin_stop_manager.php:26]: done

File: application/plugins/stop_manager/media/mysql_stop_manager.sql (hunk 3 above)
Comment (tenzap) [reply application/plugins/stop_manager/media/mysql_stop_manager.sql]: done

**Comment by tenzap** [comment]
Changing the name implies to rename almost everything (with the risks that go with it). Shall we really do this ?

**Comment by kingster** [comment]
> Changing the name implies to rename almost everything (with the risks that go with it). Shall we really do this ?

Didn't get you. What do you mean by renaming? I meant to say we rename "stop_manager" as "Subscription Manager"

**Comment by tenzap** [comment]
Oh you meant just changing the name displayed to the user (ie the label of the plugin) ?

Because I thought you wanted to change all filenames, classes, directories, tables and method that contain "stop_manager"

**Comment by kingster** [comment]
Ideally, folder_name / table_name should match the plugin name. 

Also until the plugin is merged into the repo I don't see any risk :) What kind of risk do you forsee?

**Comment by tenzap** [comment]
if it's just the main folder name & table, the risk is not too big
changing all method names (there is almost everywhere stop_manager) would require to do all tests again since there could be easily anything missed or changed while it shouldn't have been.

**Comment by tenzap** [comment]
Just as a reminder, the table doesn't save subscriptions, but only the "unsubscriptions" ie the numbers that have asked to opt-out.
Are you still sure to want it renamed to "Subscription Manager" ?

**Comment by kingster** [comment]
> Just as a reminder, the table doesn't save subscriptions, but only the "unsubscriptions" ie the numbers that have asked to opt-out.

Well true I wanted a single "Subscription Manager" ideally which could manage both subscriptions and unsubscriptions.  If you think that is not possible, then probably "Stop Manager" is good enough for now.

Though for tests you would need to move them. Have a look at #227 (You would need to move them to tests directory.

**Comment by tenzap** [comment]
Requested changes have been completed. Please review again for merge :)

**Comment by kingster** [comment]
Looks much better now.

**Comment by kingster** [comment]
Please add the test-cases now as #227 is merged now.

**Comment by tenzap** [comment]
Looks like the 2 things I called test don't really match the meaning of tests in ci-phpunit-test (and probably not the meaning of testing in software development)

The two functions I had were just here to call 2 methods that are actually not in classes and to check if they work. They didn't do any testing in the sense of unit test, the purpose was simply to simulate incoming & outgoing SMS (because I didn't have a real device connected)

**Pull request merged**